python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,741 @@
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    Singleton,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+JAVA_MAX_INT = 2147483647
+JAVA_MIN_INT = -2147483648
+JAVA_MAX_FLOAT = 3.4028235e38
+JAVA_MIN_FLOAT = -3.4028235e38
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+EPOCH_DAY = EPOCH.date()
+
+"""
+Iceberg literal is wrapper class used in expressions, which return unbound predicates
+It's being organized as below
+Literal
+|-- AboveMax
+|-- BelowMin
+|-- BaseLiteral
+    |-- StringLiteral
+    |-- FixedLiteral
+    |-- BinaryLiteral
+    |-- ComparableLiteral
+        |-- BooleanLiteral
+        |-- IntegerLiteral
+        |-- LongLiteral
+        |-- FloatLiteral
+        |-- DoubleLiteral
+        |-- DateLiteral
+        |-- TimeLiteral
+        |-- TimestampLiteral
+        |-- DecimalLiteral
+        |-- UUIDLiteral
+"""
+
+
+class Literal:
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def to_byte_buffer(self):
+        raise NotImplementedError()
+
+
+@singledispatch
+def of(value):
+    """
+    A generic Literal factory to construct an iceberg Literal based on python primitive data type
+    using dynamic overloading
+
+    Args:
+        value(python primitive type): the value to be associated with literal
+
+    Example:
+        import iceberg.expressions.literals
+        >>> iceberg.expressions.literals.of(1)
+        IntegerLiteral(1)
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {value}")
+
+
+@of.register(bool)
+def _of(value):
+    return BooleanLiteral(value)
+
+
+@of.register(int)  # type: ignore[no-redef]
+def _of(value):
+    """
+    Upgrade to long if python int is outside the JAVA_MIN_INT and JAVA_MAX_INT
+    """
+    if value < JAVA_MIN_INT or value > JAVA_MAX_INT:
+        return LongLiteral(value)
+    return IntegerLiteral(value)
+
+
+@of.register(float)  # type: ignore[no-redef]
+def _of(value):
+    """
+    Upgrade to double if python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
+    """
+    if value < JAVA_MIN_FLOAT or value > JAVA_MAX_FLOAT:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
+
+
+@of.register(str)  # type: ignore[no-redef]
+def _of(value):
+    return StringLiteral(value)
+
+
+@of.register(uuid.UUID)  # type: ignore[no-redef]
+def _of(value):
+    return UUIDLiteral(value)
+
+
+@of.register(bytes)  # type: ignore[no-redef]
+def _of(value):
+    return FixedLiteral(value)
+
+
+@of.register(bytearray)  # type: ignore[no-redef]
+def _of(value):
+    return BinaryLiteral(value)
+
+
+@of.register(Decimal)  # type: ignore[no-redef]
+def _of(value):
+    return DecimalLiteral(value)
+
+
+class BaseLiteral(Literal):
+    """Base literal which has a value and can be converted between types"""
+
+    def __init__(self, repr_string: str, value):
```

Comment:
type annotation for `value`?

Reply:
for `value` here it would be typed of `Any` as well, let me know if you want me to add it